### ISSUE TITLE
vim mode: Pressing enter in the buffer search dialog moves to the second match

### ISSUE THREAD

**Issue by hferreiro** [zed-industries/zed#8049]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Search for a common string that appears multiple times on the current buffer.
2. Press Enter.
3. The cursor position is changed to the second match.

I'd expect that the cursor is positioned at the first match.

### Environment

Zed: v0.122.2 (Zed)
OS: macOS 14.3.1
Memory: 32 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by ConradIrwin** [comment]
To clarify reproduction steps here:

* have vim mode enabled
* use `cmd-f` to open search
* search for something with multiple results
* hit `enter`

(This doesn't happen if you use `/` to open search).

See also #7692

**Comment by ConradIrwin** [comment]
Hmm. I definitely did, thanks for clarifying.

I now can't reproduce this at all. Can you describe the reproduction steps in more detail?

**Comment by hferreiro** [comment]
1. press `/` to enter the search dialog after launching Zed
2. write a word which appears multiple times in the buffer (e.g. "test")
3. press enter. Now the cursor is at the first occurrence of "test"
5. press `/`
6. press enter. Now the cursor is at the second occurrence of "test"

**Comment by d1y** [comment]
I think it's the same problem: https://github.com/zed-industries/zed/issues/7692

**Comment by ConradIrwin** [comment]
@hferreiro hmm... The behavior you are describing is how it should work (each time you do `/<enter>` you should be taken to the next match).

I think I managed to reproduce what you're seeing using just vim mode though:

* Open README.md
* Type `/ o u enter` (it should match the "You" on line 9)
* Click on line 8
* Type `/ o u enter` again (re-typing the search).
* It jumps to the "our" on line 11 instead of going back to the "You" on line 9.

https://github.com/zed-industries/zed/assets/94272/c32964aa-f32c-49c8-b842-66ba2d3bf624

**Comment by hferreiro** [comment]
Yes, but that's not the correct behavior. Enter should just unfocus and jump to the first result, as mentioned at https://github.com/zed-industries/zed/issues/7692#issuecomment-1941912495.

**Comment by hferreiro** [comment]
Funny thing, at the forth step at https://github.com/zed-industries/zed/issues/8049#issuecomment-1955897786, if you don't type `o` `u` ( because the search string already appears in the dialog) and just press `enter`, it works as expected.

**Comment by hferreiro** [comment]
Still an issue.

**Comment by hferreiro** [comment]
Still an issue.

**Comment by nilehmann** [comment]
The issue is caused by this `if`:

https://github.com/zed-industries/zed/blob/74d1a652f6404f0a5faff4ad1795d8f329107943/crates/vim/src/normal/search.rs#L191

That `if` skips the first match if the query hasn't changed, that's why this is triggered when you search the same thing twice. Removing that line fixes the issue for me but some tests fail. I don't know exactly what the logic there is, but I think it's trying to emulate the behavior in vim when the current selection overlaps with a match, if this happens vim skips the match under the cursor and goes to the next. This is however not correctly replicated as evidenced by:

initial state: `«ˇa»a    aa`
keystrokes: `/ a a enter`
final state: 
* zed:  `«ˇaa»    aa`
* vim: `«aa    ˇaa»`

**Comment by ConradIrwin** [comment]
:+1: Nice find! I'd like to fix this to more correctly match vim, and it sounds like we need to maintain a little more or different state to correctly emulate how vim works here.

Happy to pair on a solution, or review PRs if you have an idea: https://cal.com/conradirwin/pairing

**Comment by nilehmann** [comment]
I took a slot for Tuesday
